picoqinq/qinqclient.c

```diff
@@ -0,0 +1,451 @@
+uint8_t* picoquic_frames_varint_decode(uint8_t* bytes, const uint8_t* bytes_max, uint64_t* n64);
+uint8_t* picoquic_frames_varlen_decode(uint8_t* bytes, const uint8_t* bytes_max, size_t* n);
+uint8_t* picoquic_frames_uint8_decode(uint8_t* bytes, const uint8_t* bytes_max, uint8_t* n);
+uint8_t* picoquic_frames_uint16_decode(uint8_t* bytes, const uint8_t* bytes_max, uint16_t* n);
+uint8_t* picoquic_frames_uint32_decode(uint8_t* bytes, const uint8_t* bytes_max, uint32_t* n);
+uint8_t* picoquic_frames_uint64_decode(uint8_t* bytes, const uint8_t* bytes_max, uint64_t* n);
+uint8_t* picoquic_frames_cid_decode(uint8_t* bytes, const uint8_t* bytes_max, picoquic_connection_id_t* n);
+
+
+/* Per stream context.
+ *
+ * QINQ uses streams to send protocol elements such as registration of connection IDs.
+ * The "end of stream" marks the end of a protocol element. When it is received, the
+ * protocol machine is called, and the response is posted on the stream.
+ */
+
+picoqinq_client_stream_ctx_t* picoqinq_find_or_create_client_stream(picoquic_cnx_t* cnx, uint64_t stream_id, picoqinq_client_callback_ctx_t* ctx, int should_create)
+{
+    picoqinq_client_stream_ctx_t* stream_ctx = NULL;
+
+    /* if stream is already present, check its state. New bytes? */
+    stream_ctx = ctx->first_stream;
+    while (stream_ctx != NULL && stream_ctx->stream_id != stream_id) {
+        stream_ctx = stream_ctx->next_stream;
+    }
+
+    if (stream_ctx == NULL && should_create) {
+        stream_ctx = (picoqinq_client_stream_ctx_t*)
+            malloc(sizeof(picoqinq_client_stream_ctx_t));
+        if (stream_ctx == NULL) {
+            /* Could not handle this stream */
+            picoquic_reset_stream(cnx, stream_id, PICOQINQ_ERROR_INTERNAL);
+        }
+        else {
+            memset(stream_ctx, 0, sizeof(picoqinq_client_stream_ctx_t));
+            stream_ctx->next_stream = ctx->first_stream;
+            ctx->first_stream = stream_ctx;
+            stream_ctx->stream_id = stream_id;
+        }
+    }
+
+    return stream_ctx;
+}
+
+
+void picoqinq_forget_client_stream(picoqinq_client_callback_ctx_t* ctx, picoqinq_client_stream_ctx_t* stream_ctx)
+{
+    if (ctx != NULL && stream_ctx != NULL) {
+        picoqinq_client_stream_ctx_t** previous_link = &ctx->first_stream;
+        while (*previous_link != NULL && *previous_link != stream_ctx) {
+            previous_link = &(*previous_link)->next_stream;
+        }
+        if (*previous_link == stream_ctx) {
+            *previous_link = stream_ctx->next_stream;
+        }
+        if (stream_ctx->hc != NULL) {
+            free(stream_ctx->hc);
+            stream_ctx->hc = NULL;
+        }
+        free(stream_ctx);
+    }
+}
+
+/* Management of header compression
+ */
+
+int picoqinq_client_is_register_address_pending(picoqinq_client_callback_ctx_t* ctx, struct sockaddr* addr, picoquic_connection_id_t* cid, uint64_t direction)
```

Comment:
All passed pointers should be const here.

Reply:
OK. I fixed that here and in a set of other places.